File: BookingClient/src/app/model/Room.ts
export class Room {
  // @ts-ignore
  id: number;
  // @ts-ignore
  name: string;
  // @ts-ignore
  location: string;
  capacities: Array<LayoutCapacity>;

  constructor(name?: string, location?: string, id?: number) {
    if (name) {
      this.name = name;
    }
    if (location) {
      this.location = location;
    }
    if (id) {
      this.id = id;
    }
    this.capacities = new Array<LayoutCapacity>();
  }

  addLayoutCapacity(layoutCapacity: LayoutCapacity): void {
    this.capacities.push(layoutCapacity);
  }
}

export class LayoutCapacity {
  layout: Layout;
  capacity: number

  constructor(layout: Layout, capacity: number) {
    this.layout = layout;
    this.capacity = capacity;
  }
}

export enum Layout {
  THEATER = 'Theater',
  USHAPE = 'U-Shape',
  BOARD = 'Board Meeting'
}

File: booking/src/main/java/learn/ang/booking/model/entities/Room.java
package learn.ang.booking.model.entities;

import learn.ang.booking.model.Layout;
import lombok.Getter;
import lombok.Setter;

import javax.persistence.*;
import javax.validation.constraints.NotBlank;
import java.util.ArrayList;
import java.util.List;

@Entity
@Getter
@Setter
public class Room {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    @NotBlank(message = "Name cannot be blank")
    private String name;

    @NotBlank(message = "location cannot be blank")
    private String location;

    @OneToMany(cascade = {CascadeType.ALL})
    private List<LayoutCapacity> capacities;

    public Room() {
    }

    public Room(String name, String location) {
        this.name = name;
        this.location = location;
        capacities = new ArrayList<>();
        for (Layout layout : Layout.values()) {
            capacities.add(new LayoutCapacity(layout, 0));
        }
    }

    public void setCapacity(LayoutCapacity capacity) {
        for (LayoutCapacity lc : capacities) {
            if (lc.getLayout() == capacity.getLayout()) {
                lc.setCapacity(capacity.getCapacity());
            }
        }
    }
}
